feat: add pre-emptive async oauth token refreshes [draft]

igorbernstein2 opened the pull request:
This is currently a rough sketch and should not be merged. I just wanted to get some feedback here.

The current implementation of oauth refresh offloads the IO to a separate executor, however when a token expires, all calls to getRequestMetadata would hang until the token is refreshed.
This PR is a rough sketch to improve the situation by adding a stale state to token. If a token is within a minute of expiration, the first request to notice this, will spawn a refresh on the executor and immediately return with the existing token. This avoids hourly latency spikes for grpc.

The implementation uses guava's ListenableFutures to manage the async refresh state. Although the apis are marked BetaApi in guava 19, they are GA in guava 30

* The class introduces 3 distinct states:
  * Expired -  the token is not usable
  * Stale - the token is useable, but its time to refresh
  * Fresh - token can be used without any extra effort
* All of the funtionality of getRequestMetadata has been extracted to asyncFetch. The new function will check if the token is unfresh and schedule a refresh using the executor
* asyncFetch uses ListenableFutures to wrap state: if the token is not expired, an immediate future is returned. If the token is expired the future of the refresh task is returned
* A helper refreshAsync 	& finishRefreshAsync are also introduced. They schedule the actual refresh and propagate the side effects
* To handle blocking invocation: the async functionality is re-used but a DirectExecutor is used. All ExecutionErrors are unwrapped. In most cases the stack trace will be preserved because of the DirectExecutor. However if the async & sync methods are interleaved, it's possible that a sync call will await an async refresh task. In this case the callers stacktrace will not be present.

igorbernstein2 commented on oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:
```diff
@@ -208,6 +210,27 @@ private ImpersonatedCredentials initializeImpersonatedCredentials() {
         .build();
   }
 
+  @Override
```
Note to reviewer: before this PR, the blocking version getRequestMetadata() was the primary codepath, so ExternalAccountCredentials only had to override that method. This subclass expected that the async version would just call the sync version. Which I think is a poor assumption since its an implementation detail. I corrected the behavior by overriding both.

Arguably a better approach is to make both getRequestMetadata final and have subclasses use  asyncFetch() as the source of truth, but I didn't want to leak guava to the surface of this library

TimurSadykov commented on oauth2_http/java/com/google/auth/oauth2/OAuth2Credentials.java:
```diff
@@ -144,75 +152,179 @@ public void getRequestMetadata(
+        // noop
+      } finally {
+        if (this.refreshTask == finishedTask) {
+          this.refreshTask = null;
```
can we make refreshTask @VisibleForTesting and cover its reset to null this with some tests? To avoid some nasty regressions.

igorbernstein2 replied:
I updated the existing tests to ensure that its managed properly. Let me know if you had something else in mind

TimurSadykov replied:
looks good, thx!

TimurSadykov commented on oauth2_http/javatests/com/google/auth/oauth2/OAuth2CredentialsTest.java:
```diff
@@ -348,6 +397,211 @@ public void getRequestMetadata_temporaryToken_hasToken() throws IOException {
+  }
+
+  @Test
+  public void getRequestMetadata_singleFlightErrorSharing()
```
what is the meaning of the "singleFlightErrorSharing"? The test makes sense, but unfamiliar with the terminology

igorbernstein2 replied:
Internal link is go/singleflight, but it seems to exist in public domain as well. Single flight is a mechanic to coalesce multiple requests into one. A general implementation would allow for batching, but in this case we only have a single operation type that we need to combine

TimurSadykov replied:
Thanks for pointers :)

TimurSadykov reviewed (changes requested):
Overall looks good, lets clarify if refresh slight behavior change is blocking. 

Some coverage of refreshTask would be great.

igorbernstein2 commented on oauth2_http/java/com/google/auth/oauth2/OAuth2Credentials.java:
```diff
@@ -144,75 +152,179 @@ public void getRequestMetadata(
+    return unwrapDirectFuture(asyncFetch(MoreExecutors.directExecutor())).requestMetadata;
   }
 
   /** Refresh the token by discarding the cached token and metadata and requesting the new ones. */
```
yep, I updated the docs. Let me know if its more it makes more sense

TimurSadykov replied:
The comment looks good, FYI @silvolu to double check that its not a breaking change

TimurSadykov replied:
@igorbernstein2 i synced with @silvolu, there is no concern with this change.

igorbernstein2 reviewed (commented):
Thanks for reviewing! I think addressed all of your comments, PTAL

paulmordont commented:
Hey @igorbernstein2 ! Thx a lot for the PR, from my side everything looks good. Looking forward to testing this in the next bigtable client release :)

paulmordont commented on oauth2_http/java/com/google/auth/oauth2/OAuth2Credentials.java:
```diff
@@ -144,75 +152,182 @@ public void getRequestMetadata(
+    AsyncRefreshResult refreshResult = null;
+
+    // Schedule a refresh as necessary
     synchronized (lock) {
```
I wonder Is it still necessary to go inside the lock if the token state `Fresh`? Is it viable to just do the following:
```
if (getState() == CacheState.Fresh) {
    return Futures.immediateFuture(value);
}
```

igorbernstein2 replied:
Thats good observation, I added the optimization. Thanks!

TimurSadykov reviewed (commented):
Looks good to me

igorbernstein2 commented:
I created cl/373364841, I had to hand edit a couple of irrelevant files:
- added missing RunWith on unrelated tests
- added missing deps in BUILD

I requested a global tap presubmit

Also, I updated the codestyle suggestions from critique

lsirac commented on oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:
```diff
@@ -208,6 +210,27 @@ private ImpersonatedCredentials initializeImpersonatedCredentials() {
+        new RequestMetadataCallback() {
+          @Override
+          public void onSuccess(Map<String, List<String>> metadata) {
+            metadata = addQuotaProjectIdToRequestMetadata(quotaProjectId, metadata);
```
do we need to do this here? the quotaProjectId is being added in the override below, which seems like is always called

igorbernstein2 replied:
This doesn't call the override below. It calls `super.getRequestMetadata`

igorbernstein2 replied:
But I did just notice that I had a typo & invoked super.getRequestMetadata twice.

lsirac replied:
ah calling super.getRequestMetadata used to call the override below

igorbernstein2 replied:
yea, I dont love the parallel logic paths, and tried to call it out in the comment above, but I dont have a better approach. If you have ideas, lets keep the discussion in https://github.com/googleapis/google-auth-library-java/pull/646/files#r625406829
